cli/azd/pkg/infra/provisioning/bicep_provider.go

```diff
@@ -75,203 +107,345 @@ func (p *BicepProvider) Preview(ctx context.Context) *async.InteractiveTaskWithP
 					param.Value = bicepParam.Value
-					template.Parameters[key] = param
+					deployment.Parameters[key] = param
+				}
+			}
+
+			updated, err := p.ensureParameters(ctx, deployment)
+			if err != nil {
+				asyncContext.SetError(err)
+				return
+			}
+
+			if updated {
+				if err := p.updateParametersFile(ctx, deployment); err != nil {
+					asyncContext.SetError(fmt.Errorf("updating deployment parameters: %w", err))
+					return
 				}
 			}
 
 			result := PreviewResult{
-				Preview: *template,
+				Deployment: *deployment,
 			}
 
 			asyncContext.SetResult(&result)
 		})
 }
 
-func (p *BicepProvider) UpdatePlan(ctx context.Context, preview Preview) error {
-	bicepFile := BicepTemplate{
-		Schema:         "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
-		ContentVersion: "1.0.0.0",
-	}
+// Provisioning the infrastructure within the specified template
+func (p *BicepProvider) Deploy(ctx context.Context, deployment *Deployment, scope Scope) *async.InteractiveTaskWithProgress[*DeployResult, *DeployProgress] {
+	return async.RunInteractiveTaskWithProgress(
+		func(asyncContext *async.InteractiveTaskContextWithProgress[*DeployResult, *DeployProgress]) {
+			done := make(chan bool)
+			var operations []azcli.AzCliResourceOperation
 
-	parameters := make(map[string]BicepInputParameter)
+			deploymentSlug := azure.SubscriptionDeploymentRID(p.env.GetSubscriptionId(), p.env.GetEnvName())
+			deploymentUrl := fmt.Sprintf(output.WithLinkFormat("https://portal.azure.com/#blade/HubsExtension/DeploymentDetailsBlade/overview/id/%s\n"), url.PathEscape(deploymentSlug))
+			p.console.Message(ctx, fmt.Sprintf("Provisioning Azure resources can take some time.\n\nYou can view detailed progress in the Azure Portal:\n%s", deploymentUrl))
 
-	for key, param := range preview.Parameters {
-		parameters[key] = BicepInputParameter(param)
-	}
+			// Report incremental progress
+			go func() {
+				resourceManager := infra.NewAzureResourceManager(ctx)
+				progressDisplay := NewProvisioningProgressDisplay(resourceManager, p.console, p.env.GetSubscriptionId(), p.env.GetEnvName())
+
+				for {
+					select {
+					case <-done:
+						return
+					case <-time.After(10 * time.Second):
+						progressReport, err := progressDisplay.ReportProgress(ctx, asyncContext)
+						if err != nil {
+							asyncContext.SetError(err)
+							return
+						}
+
+						operations = progressReport.Operations
+						asyncContext.SetProgress(progressReport)
+					}
+				}
+			}()
 
-	bicepFile.Parameters = parameters
+			// Start the deployment
+			modulePath := p.modulePath()
+			parametersFilePath := p.parametersFilePath()
+			deployResult, err := p.deployModule(ctx, scope, modulePath, parametersFilePath)
 
-	bytes, err := json.MarshalIndent(bicepFile, "", "  ")
-	if err != nil {
-		return fmt.Errorf("marshaling parameters: %w", err)
-	}
+			if err != nil {
+				asyncContext.SetError(err)
```

Comment:
Fixed to ensure `done<-true` is always run.

Reply:
For folks wondering, since Wei made this comment, Wallace added a defer which sends the done message so we send it in all cases, but GitHub was not smart enough to notice.

Reply:
@ellismg @wbreza We should probably check what the dotnet implementation is if multiple `SetResult` or `SetError` is called. I suspect we don't want to panic and bring down the application in case it does happen....

Reply:
Nvm, Wallce had addressed it :) Thanks.